File: elixir/captains-log/lib/captains_log.ex
defmodule CaptainsLog do
  @planetary_classes ["D", "H", "J", "K", "L", "M", "N", "R", "T", "Y"]

  def random_planet_class() do
    Enum.random(@planetary_classes)
  end

  def random_ship_registry_number() do
    random_num = :rand.uniform(9000) + 999
    "NCC-#{random_num}"
  end

  def random_stardate() do
    41000.0 + :rand.uniform() * 1000.0
  end

  def format_stardate(stardate) when is_float(stardate),
    do: Float.round(stardate, 1) |> Float.to_string()

  def format_stardate(stardate), do: :io_lib.format("~.1f", [stardate]) |> to_string()
end

File: java/captains-log/src/main/java/CaptainsLog.java
import java.util.Random;

class CaptainsLog {

  private static final char[] PLANET_CLASSES = new char[] { 'D', 'H', 'J', 'K', 'L', 'M', 'N', 'R', 'T', 'Y' };

  private Random random;

  CaptainsLog(Random random) {
    this.random = random;
  }

  char randomPlanetClass() {
    final int randomNumber = this.random.nextInt(PLANET_CLASSES.length);

    return PLANET_CLASSES[randomNumber];
  }

  String randomShipRegistryNumber() {

    return "NCC-" + (1000 + this.random.nextInt(9000));
  }

  double randomStardate() {
    return 41000 + 1000 * this.random.nextDouble();
  }
}
